File: casciffo-front-end/src/model/research/ResearchModel.ts
import {ProposalModel} from "../proposal/ProposalModel";
import {StateTransitionModel} from "../state/StateTransitionModel";
import {StateModel} from "../state/StateModel";
import UserModel from "../user/UserModel";
import {TeamInvestigatorModel} from "../user/TeamInvestigatorModel";
import {FileInfo} from "../proposal/finance/ProposalFinanceModel";


export interface ResearchModel {
    id?: string
    lastModified?: string
    eudra_ct?: string
    sampleSize?: string
    duration?: string
    cro?: string
    startDate?: string
    endDate?: string
    estimatedEndDate?: string
    estimatedPatientPool?: string
    actualPatientPool?: string
    industry?: string
    protocol?: string
    initiativeBy?: string
    phase?: string
    type?: string
    treatmentType?: string
    typology?: string
    specification?: string
    treatmentBranches?: string // comes separated by ';'
    stateId?: string
    state?: StateModel
    proposalId?: string
    proposal?: ProposalModel
    visits?: ResearchVisitModel[]
    addendas?: ResearchAddenda[]
    stateTransitions?: StateTransitionModel[]
    dossiers?: DossierModel[]
    patients?: ResearchPatientModel[]
    scientificActivities?: ScientificActivityModel[]
    investigationTeam?: TeamInvestigatorModel[]
    financeComponent?: ResearchFinance
    canceledReason?: string
    canceledById?: string
    canceledBy?: UserModel
}

export interface ResearchAggregateModel {
    id?: string
    eudra_ct?: string
    sampleSize?: string
    lastModified?: string
    duration?: string
    cro?: string
    startDate?: string
    endDate?: string
    estimatedEndDate?: string
    industry?: string
    protocol?: string
    initiativeBy?: string
    phase?: string
    type?: string
    stateId?: string
    stateName?: string
    proposalId?: string
    sigla?: string
    serviceName?: string
    serviceId?: string
    therapeuticAreaName?: string
    therapeuticAreaId?: string
    pathologyName?: string
    pathologyId?: string
    principalInvestigatorId?: string
    principalInvestigatorName?: string
    principalInvestigatorEmail?: string
    promoterName?: string
    hasPartnerships?: boolean
}

export interface DossierModel {
    id?: string
    clinicalResearchId?: string
    volume?: string
    label?: string
    amount?: number
}

export interface ResearchVisitModel {
    id?: string
    researchId?: string
    researchPatientId?: string
    concluded?: boolean;
    visitType?: string
    scheduledDate?: string
    startDate?: string
    endDate?: string
    periodicity?: string
    customPeriodicity?: number
    observations?: string
    hasAdverseEventAlert?: boolean
    hasMarkedAttendance?: boolean
    researchPatient?: ResearchPatientModel
    visitInvestigators?: VisitInvestigator[]
}

export interface VisitInvestigator {
    id?: string
    visitId?: string
    investigatorId?: string
    investigator?: UserModel
}

export interface PatientModel {
    id?: string
    processId?: string
    fullName?: string
    gender?: string
    age?: string
}

export interface ResearchPatientModel {
    id?: string
    patientId?: string,
    researchId?: string,
    joinDate?: string
    lastVisitDate?: string
    treatmentBranch?: string,
    patient?: PatientModel
}

export interface PatientVisitsAggregate {
    patient: PatientModel
    scheduledVisits: ResearchVisitModel[]
}

export interface ResearchPatientVisitsAggregate {
    researchPatient: ResearchPatientModel
    scheduledVisits: ResearchVisitModel[]
}

export interface ScientificActivityModel {
    id?: string
    researchId?: string
    datePublished?: string
    author?: string
    paperName?: string
    volume?: string
    volumeNumber?: string
    paperNumPages?: string
    countryPublished?: string
    hasBeenIndexed?: boolean
    publishedUrl?: string
    publicationType?: string
    researchType?: string
}

export interface ResearchAddenda {
    id?: string
    researchId?: string
    stateId?: string
    fileId?: string
    createdDate?: string
    state?: StateModel
    stateTransitions?: StateTransitionModel[]
    observations?: AddendaCommentsModel[]
    fileInfo?: FileInfo
}

export interface AddendaCommentsModel {
    id?: string
    createdDate?: string
    addendaId?: string
    observation?: string
    authorId?: string
    authorName?: string
    author?: UserModel
}

export interface ResearchModelAnswer {
    research?: ResearchModel
    success?: boolean
}

export interface ResearchFinance {
    id?: string
    researchId?: string
    valuePerParticipant?: string
    roleValuePerParticipant?: string
    balance?: string
    teamFinanceFlow?: ResearchTeamFinanceEntry[]
    monetaryFlow?: ResearchFinanceEntry[]
}

export interface ResearchFinanceEntry {
    id?: string
    rfcId?: string
    transactionDate?: string
    typeOfMonetaryFlow?: string
    motive?: string
    amount?: string
}

export interface ResearchTeamFinanceEntry {
    id?: string
    rfcId?: string
    typeOfMonetaryFlow?: string
    transactionDate?: string
    responsibleForPayment?: string
    amount?: string
    partitionPercentage?: string
    roleAmount?: string
    investigatorId?: string
    investigator?: UserModel
}

export interface ResearchFinanceNewEntryDTO {
    id?: string
    researchId?: string
    valuePerParticipant?: string
    roleValuePerParticipant?: string
    balance?: string
    newMonetaryEntry?: ResearchFinanceEntry
    newTeamFinanceEnty?: ResearchTeamFinanceEntry
}

File: casciffo-spring-backend/src/main/kotlin/isel/casciffo/casciffospringbackend/research/research/ResearchModel.kt
package isel.casciffo.casciffospringbackend.research.research

import isel.casciffo.casciffospringbackend.common.ResearchType
import isel.casciffo.casciffospringbackend.investigation_team.InvestigationTeamModel
import isel.casciffo.casciffospringbackend.proposals.proposal.ProposalModel
import isel.casciffo.casciffospringbackend.research.addenda.Addenda
import isel.casciffo.casciffospringbackend.research.dossier.Dossier
import isel.casciffo.casciffospringbackend.research.finance.overview.ResearchFinance
import isel.casciffo.casciffospringbackend.research.patients.ResearchPatient
import isel.casciffo.casciffospringbackend.research.studies.ScientificActivity
import isel.casciffo.casciffospringbackend.research.visits.visits.VisitModel
import isel.casciffo.casciffospringbackend.states.state.State
import isel.casciffo.casciffospringbackend.states.transitions.StateTransition
import isel.casciffo.casciffospringbackend.users.user.UserModel
import kotlinx.coroutines.flow.Flow
import org.springframework.beans.factory.annotation.Value
import org.springframework.data.annotation.Id
import org.springframework.data.annotation.Transient
import org.springframework.data.relational.core.mapping.Column
import org.springframework.data.relational.core.mapping.Table
import java.time.LocalDate
import java.time.LocalDateTime

@Table("clinical_research")
data class ResearchModel (
    @Id
    @Column("research_id")
    var id: Int? = null,
    var proposalId: Int? = null,
    var lastModified: LocalDateTime? = null,
    @Column("research_state_id")
    var stateId: Int? = null,
    var eudra_ct: String? = null,
    var sampleSize: Int? = null,
    var duration: Int? = null,
    var cro: String? = null,
    var startDate: LocalDate? = null,
    var endDate: LocalDate? = null,
    var estimatedEndDate: LocalDate? = null,
    var estimatedPatientPool: Int? = null,
    var actualPatientPool: Int? = null,
    var industry: String? = null,
    var protocol: String? = null,
    var initiativeBy: String? = null,
    var phase: String? = null,
    var type: ResearchType? = null,
    var treatmentType: String? = null,
    var typology: String? = null,
    var specification: String? = null,
    var treatmentBranches: String? = null,
    var canceledReason: String? = null,
    var canceledById: Int? = null,

    @Transient
    @Value("null")
    var canceledBy: UserModel? = null,

    @Transient
    @Value("null")
    var state: State? = null,

    @Transient
    @Value("null")
    var proposal: ProposalModel? = null,

    @Transient
    @Value("null")
    var visits: Flow<VisitModel>? = null,

    @Transient
    @Value("null")
    var patients: Flow<ResearchPatient>? = null,

    @Transient
    @Value("null")
    var dossiers: Flow<Dossier>? = null,

    @Transient
    @Value("null")
    var stateTransitions: Flow<StateTransition>? = null,

    @Transient
    @Value("null")
    var scientificActivities: Flow<ScientificActivity>? = null,

    @Transient
    @Value("null")
    var investigationTeam: Flow<InvestigationTeamModel>? = null,

    @Transient
    @Value("null")
    var addendas: Flow<Addenda>? = null,

    @Transient
    @Value("null")
    var financeComponent: ResearchFinance? = null
)